cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -0,0 +1,335 @@
+      if (failureCount > 0) {
+        // Exponential strategy doesn't use the time component, so doesn't matter what we passed in as startTime
+        return scheduleStrategy.nextRetry(failureCount, 0);
+      }
+
+      // Sleep for 2 seconds if there's no jobs in the queue
+      return emptyFetch ? 2000L : 0L;
+    }
+
+    private boolean checkJobConstraints(DatasetContext context) throws Exception {
+      boolean emptyScan = true;
+
+      JobQueueDataset jobQueue = getJobQueue(context);
+      CloseableIterator<Job> jobQueueIter;
+      if (lastConsumed == null) {
+        jobQueueIter = jobQueue.getJobs(partition);
+      } else {
+        jobQueueIter = jobQueue.getJobs(partition, lastConsumed);
+      }
+
+      try {
+        Stopwatch stopWatch = new Stopwatch().start();
+        // limit the batches of the scan to 1000ms
+        while (isRunning() && stopWatch.elapsedMillis() < 1000) {
+          if (!jobQueueIter.hasNext()) {
+            jobQueueIter.close();
+            lastConsumed = null;
+            return emptyScan;
+          }
+          Job job = jobQueueIter.next();
+          lastConsumed = job;
+          emptyScan = false;
+          checkAndUpdateJob(jobQueue, job);
+        }
+      } finally {
+        jobQueueIter.close();
+      }
+      return emptyScan;
+    }
+
+    private void checkAndUpdateJob(JobQueueDataset jobQueue, Job job) {
+      switch (job.getState()) {
+        case PENDING_TRIGGER:
+          if (isTriggerSatisfied(job)) {
+            jobQueue.updateState(job, Job.State.PENDING_CONSTRAINT);
+          }
+          break;
+        case PENDING_CONSTRAINT:
+        case PENDING_LAUNCH:
+          if (constraintsSatisfied(job)) {
+            jobQueue.updateState(job, Job.State.PENDING_LAUNCH);
+            readyJobs.add(job);
+          }
+          break;
+      }
+    }
+
+    private boolean runReadyJobs(DatasetContext context) throws IOException, DatasetManagementException {
+      JobQueueDataset jobQueue = getJobQueue(context);
+      Iterator<Job> readyJobsIter = readyJobs.iterator();
+      while (readyJobsIter.hasNext() && isRunning()) {
+        Job job = readyJobsIter.next();
+        readyJobsIter.remove();
+
+        // We should check the stored job's state (whether it actually is PENDING_LAUNCH), because
+        // the schedule could have gotten deleted in the meantime or the transaction that marked it as PENDING_LAUNCH
+        // may have failed / rolled back.
+        Job storedJob = jobQueue.getJob(job.getJobKey());
+        if (storedJob == null) {
+          continue;
+        }
+
+        if (storedJob.getState() == Job.State.PENDING_LAUNCH) {
+          ProgramSchedule schedule = job.getSchedule();
+          try {
+            // TODO: Temporarily execute scheduled program without any checks. Need to check appSpec, scheduleSpec
+            taskRunner.execute(schedule.getProgramId(), ImmutableMap.<String, String>of(),
```

Comment:
Can get userArgs and sysArgs from `propertiesResolver` and properties in schedule